bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java

```diff
@@ -173,25 +182,38 @@ public CloudBigtableScanConfiguration build() {
    * @param request The {@link ReadRowsRequest} that will be used to filter the table.
    * @param additionalConfiguration A {@link Map} with additional connection configuration.
    */
-  protected CloudBigtableScanConfiguration(String projectId, String instanceId, String tableId,
-      ReadRowsRequest request, Map<String, String> additionalConfiguration) {
-    super(projectId, instanceId,  tableId, additionalConfiguration);
-    if (request.getTableName().isEmpty()) {
-      BigtableInstanceName bigtableInstanceName =
-          new BigtableInstanceName(projectId, this.getInstanceId());
-      String fullTableName = bigtableInstanceName.toTableNameStr(tableId);
-      this.request = request.toBuilder().setTableName(fullTableName).build();
-    } else {
-      this.request = request;
-    }
+  protected CloudBigtableScanConfiguration(
+      ValueProvider<String> projectId,
+      ValueProvider<String> instanceId,
+      String tableId,
+      ReadRowsRequest request,
+      Map<String, ValueProvider<String>> additionalConfiguration) {
+    super(projectId, instanceId, tableId, additionalConfiguration);
+    this.request =
+        NestedValueProvider.of(
+            // Eventually the input request will be ValueProvider<ReadRowsRequest>.
+            StaticValueProvider.of(request),
+            new SerializableFunction<ReadRowsRequest, ReadRowsRequest>() {
+              @Override
+              public ReadRowsRequest apply(ReadRowsRequest request) {
+                if (!request.getTableName().isEmpty()) {
+                  return request;
+                }
+
+                BigtableInstanceName bigtableInstanceName =
+                    new BigtableInstanceName(getProjectId(), getInstanceId());
+                String fullTableName = bigtableInstanceName.toTableNameStr(getTableId());
+                return request.toBuilder().setTableName(fullTableName).build();
+              }
+            });
```

Comment:
I dont think this is right...`this.request` will end up with an incomplete `ValueProvider` that will return `true` for `isAccessible` when projectId or instanceId are not accessible. I think you need to create a your own subclass of `ValueProvider` that wraps all of the relevant `ValueProvider`s and in its implementation `isAccessible` it takes an && of all its component ValueProviders.

Reply:
I don't think this matters at the moment, no? If you use CloudBigtableScanConfiguration now, there is no way you can set runtime parameters, therefore isAccessible is a non-issue.

Reply:
The request object can't be modified before wrapping it in StaticValueProvider. I added TODO.

Reply:
SGTM